gadopt/time_stepper.py

```diff
@@ -117,120 +124,189 @@ def advance(
         self.get_final_solution()
 
 
-class ERKGeneric(RungeKuttaTimeIntegrator):
-    """Generic explicit Runge-Kutta time integrator.
+class IrksomeIntegrator(TimeIntegratorBase):
+    """Time integrator using Irksome as the backend.
 
-    Implements the Butcher form. All terms in the equation are treated explicitly.
-
-    Arguments:
-      equation: G-ADOPT equation to solve
-      solution: Firedrake function reperesenting the equation's solution
-      dt: Integration time step
-      solution_old: Firedrake function representing the equation's solution
-                      at the previous timestep
-      solver_parameters: Dictionary of solver parameters provided to PETSc
-      strong_bcs: List of Firedrake Dirichlet boundary conditions
+    This class wraps Irksome's TimeStepper while maintaining G-ADOPT's API
+    for compatibility with our existing code.
 
+    Args:
+        equation: G-ADOPT equation to integrate
+        solution: Firedrake function representing the equation's solution
+        dt: Integration time step (Firedrake Constant or float)
+        butcher: Irksome Butcher tableau (e.g., GaussLegendre, RadauIIA)
+        stage_type: Type of stage formulation (e.g., "deriv") # NB: Look into this more.
+        strong_bcs: List of Firedrake Dirichlet boundary conditions
+        solver_parameters: Dictionary of solver parameters provided to PETSc
     """
+
     def __init__(
         self,
         equation: Equation,
         solution: firedrake.Function,
         dt: float,
-        solution_old: Optional[firedrake.Function] = None,
-        solver_parameters: Optional[dict[str, Any]] = {},
+        butcher: ButcherTableau,
+        stage_type: str = "deriv",
         strong_bcs: Optional[list[firedrake.DirichletBC]] = None,
+        solver_parameters: Optional[dict[str, Any]] = None,
     ):
-        super(ERKGeneric, self).__init__(
-            equation, solution, dt, solution_old, solver_parameters, strong_bcs
+        super().__init__()
+
+        self.equation = equation
+        self.solution = solution
+        self.dt = float(dt)
+
+        # Keep reference to original dt constant for syncing
+        self.dt_reference = ensure_constant(dt)
+
+        # Create MeshConstant objects for time variables (what Irksome expects)
+        mesh = solution.ufl_domain()
+        mc = MeshConstant(mesh)
+        self.t = mc.Constant(0.0)
+        self.dt_mesh_const = mc.Constant(float(dt))  # Internal MeshConstant for Irksome
+
+        # Build the Irksome form: F = equation.mass(Dt(solution)) - equation.residual(solution)
+        # Note: equation.residual returns the negative of the left-hand contribution
+        F = equation.mass(Dt(solution)) - equation.residual(solution)
```

Comment:
Presumably this doesn't yet address the issue that started this - this still won't work if your mass term is of the form $$C(u)\frac{\partial u}{\partial t}$$. I think (but this does require changes on the outside) if we indeed have successfully replaced all timesteppers with IRKSome indeed, the mass term could be just like any other term where the _equation_ code would be responsible for inserting the `Dt` operator. Then the equation would just produce the entire residual, including the mass term